File: Smaller Sum[ 22-April-2023]/Binary Search based Approach/Solution.cpp
//   C++ CODE 


int binarySearch(int val, vector<int>& arr) {
        int low = 0, high = arr.size();
        while (low < high) {
            int mid = (low + high) / 2;
            if (arr[mid] < val) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }

    vector<long long> smallerSum(int n, vector<int>& arr) {
        vector<long long> ans(n);
        int in = 0;
        vector<long long> prefixSum(n);
        vector<int> arrcpy(arr);
        sort(arrcpy.begin(), arrcpy.end());
        prefixSum[0] = arrcpy[0];
        for (int i = 1; i < n; i++) {
            prefixSum[i] = prefixSum[i - 1] + arrcpy[i];
        }
        for (int i = 0; i < n; i++) {
            int index = binarySearch(arr[i], arrcpy);
            if (index != 0) {
                ans[in++] = prefixSum[index-1];
            } else {
                in++;
            }
        }
        return ans;
    }

File: Smaller Sum[ 22-April-2023]/Binary Search based Approach/Solution.java
// >> JAVA CODE <<


class Solution
{
    public int binarySearch(int val,int[] arr)
    {
        int low=0,high=arr.length-1;
        int ans=-1;
        while(low<=high)
        {
            int mid=(low+high)/2;
            
            if(arr[mid]<val)
            {
                ans=mid;
                low=mid+1;
            }
            else
                high=mid-1;
        }
        
            return ans;
    }
    public long[] smallerSum(int n,int arr[])
    {
        long[] ans=new long[n]; int in=0;
        long[] prefixSum=new long[n];
        int[] arrcpy=new int[n];
        
        for(int i=0;i<n;i++) arrcpy[i]=arr[i]; 
        
        Arrays.sort(arrcpy);
        
        prefixSum[0]=arrcpy[0];
        
        for(int i=1;i<n;i++) prefixSum[i]=prefixSum[i-1]+arrcpy[i];
        
        for(var val:arr)
        {
            int index=binarySearch(val,arrcpy);
            
            if(index!=-1)
                ans[in++]=prefixSum[index];
            else
                in++;
        }
        
            return ans;
    }
}